Sjj/unicodefixes

stevenjohnstone opened the pull request.

Commit:
Add another fuzz discovered corner case

strings.ToLower can output a string with fewer bytes than the
input. This leads to crashes

vanackere commented on matching/dictionary.go:
```diff
@@ -12,19 +14,21 @@ type dictionaryMatch struct {
 						I:              i,
 						J:              j,
-						Token:          password[i : j+1],
+						Token:          string(passwordRunes[i : j+1]),
```
It looks like this change will give different I,J indexes for strings containing unicode characters than before...
Is this why you had to change testdata/output.json ? If so, could we instead find a fix that doesn't change the previous meaning of I,J as indexes into the original password string ?

stevenjohnstone replied:
I _think_ that the original test output may have been using the ascii version of that character but in golang it's utf8 proper? I don't have node setup to regenerate output.json: maybe worth seeing if the change I made to gen.sh influences output.json?

What is clear is that strings.ToLower can give you a string of length less than that of the input. So indexing into pwLower with indices from iterating on password can go out of bounds. Perhaps the ASCII substrings (I'm assuming the dictionaries are probably printable ASCII) should be extracted and those converted into lower case for the purposes of comparison?

vanackere replied:
Can you try to only change line 21 in the original version from 
```
word := pwLower[i : j+1]
```
to
```
word := strings.ToLower(password[i : j+1])
```

I think this should solve the problem while keeping the previous index semantic

Commit:
Prevent crashes processing unicode strings

strings.ToLower can output strings with less bytes than
the input. Indexing pwLower using indices from ranging over
password can result in out of bounds accesses.

Commit:
Fix translate

For some unicode inputs, translate's output had more bytes than
the input which could lead to out of bounds accesses.

Commit:
Fix another case where translate increases byte count

Commit:
Add new corner case test

Commit:
Fix crash on corner case

Commit:
Avoid entering an infinite loop

vanackere commented on matching/leet.go:
```diff
@@ -52,15 +54,17 @@ func (lm l33tMatch) Matches(password string) []*match.Match {
-			res = res + v
-		} else {
-			res = res + string(s)
+	return strings.Map(func(r rune) rune {
```
Could you please add a unit test for translate in leet_test.go with the relevant cases ? I'd like to understand better why the translate function might return more bytes than the original (I think it should never be the case since we only should replace ascii runes)

stevenjohnstone replied:
I think TestsCornerCases has two examples where translate returns more bytes than the original. Could do with a comment or be broken out into a named test for translate.

stevenjohnstone replied:
A better explanation: the problems occur when you have a string which is not valid utf8. As you range over the string, the rune value returned may be 0xfffd which is encoded in 3 bytes.

See https://golang.org/ref/spec#For_statements:
>For a string value, the "range" clause iterates over the Unicode code points in the string starting at byte index 0. On successive iterations, the index value will be the index of the first byte of successive UTF-8-encoded code points in the string, and the second value, of type rune, will be the value of the corresponding code point. If the iteration encounters an invalid UTF-8 sequence, the second value will be 0xFFFD, the Unicode replacement character, and the next iteration will advance a single byte in the string.

Therefore, for each invalid byte value which isn't valid utf8, s will be 0xfffd and len(string(s)) is 3.

vanackere commented on matching/repeat.go:
```diff
@@ -16,6 +16,10 @@ func (repeatMatch) Matches(password string) []*match.Match {
 	var matches []*match.Match
 
 	lastIndex := 0
+	indices := map[int]bool{
```
While I'm sure this prevents an infinite loop, I'd like to understand better why the index is actually staying the same. Could you make a failing test in repeat_test.go so that we can maybe find a better / the proper fix ?

stevenjohnstone replied:
In TestCornerCases there are these two problematic strings:
"\x80夤\x1b\xbc\xaf\xa6m",
"77036262\xbd\xbfｿ\xef\xbfｿ\xef3233232".

The first causes the function to get stuck at index 6. The second causes it to cycle through some indices repeatedly.

Totally agree that this isn't the right fix. It does get fuzzing to go a bit further (yet to find any further issues).

vanackere commented:
(I've merged separately your fix in matching/dictionnary.go)

vanackere commented:
Thanks for finding this explanation !

It will be much easier to only focus on making the various matches work
safely on valid utf8 strings since passwords are expected to be valid
strings, therefore I just made a change to avoid evaluation and report a
score of 0 for passwords containing invalid utf8 runes.
From a quick run of the fuzzer there are still crashes but at least those
will be easier to understand / fix.

2018-03-30 23:12 GMT+02:00 Steven Johnstone <notifications@github.com>:

> *@stevenjohnstone* commented on this pull request.
> ------------------------------
>
> In matching/leet.go
> <https://github.com/trustelem/zxcvbn/pull/3#discussion_r178381819>:
>
> > @@ -52,15 +54,17 @@ func (lm l33tMatch) Matches(password string) []*match.Match {
>  }
>
>  func translate(password string, sub map[string]string) string {
> -	var res string
> -	for _, s := range password {
> -		if v, ok := sub[string(s)]; ok {
> -			res = res + v
> -		} else {
> -			res = res + string(s)
> +	return strings.Map(func(r rune) rune {
>
> A better explanation: the problems occur when you have a string which is
> not valid utf8. As you range over the string, the rune value returned may
> be 0xfffd which is encoded in 3 bytes.
>
> See https://golang.org/ref/spec#For_statements:
>
> For a string value, the "range" clause iterates over the Unicode code
> points in the string starting at byte index 0. On successive iterations,
> the index value will be the index of the first byte of successive
> UTF-8-encoded code points in the string, and the second value, of type
> rune, will be the value of the corresponding code point. If the iteration
> encounters an invalid UTF-8 sequence, the second value will be 0xFFFD, the
> Unicode replacement character, and the next iteration will advance a single
> byte in the string.
>
> Therefore, for each invalid byte value which isn't valid utf8, s will be
> 0xfffd and len(string(s)) is 3.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/trustelem/zxcvbn/pull/3#discussion_r178381819>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAw3eAulwZn2RPbqabn4G1Zk9alyrEzuks5tjp_DgaJpZM4S_TqY>
> .
>

vanackere commented:
I'm closing this pull request now since some parts were merged and restricting passwords to valid utf8 inputs looks to have fixed most (all ?) other cases. Feel free to reopen if you find more bugs !
